### PR TITLE
framework: build sui-framework at compile time

### PR DESCRIPTION
**Pull request by bmwill**

### CODE REVIEW

**Review by gdanezis — APPROVED** [review]
Ok, cool so I do not think my PR - caching genesis - is needed any more right? Since now genesis is statically written.

File: crates/sui-framework-build/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+use move_compiler::compiled_unit::{CompiledUnit, NamedCompiledModule};
+use move_core_types::{account_address::AccountAddress, ident_str, language_storage::ModuleId};
+use move_package::BuildConfig;
+use std::{collections::HashSet, path::Path};
+use sui_types::error::{SuiError, SuiResult};
+use sui_verifier::verifier as sui_bytecode_verifier;
+
+pub fn build_sui_framework_modules(lib_dir: &Path) -> SuiResult<Vec<CompiledModule>> {
+    let modules = build_framework(lib_dir)?;
+    verify_modules(&modules)?;
```
[the remark this answers is not in the document]
Comment (bmwill) [reply crates/sui-framework-build/src/lib.rs:14]: This function was moved and I don't know the purpose behind the verifying so i'll leave that for the move folks to deal with :)

File: crates/sui-framework-build/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        ident_str!("Event").to_owned(),
+        ident_str!("GUID").to_owned(),
+        #[cfg(not(test))]
+        ident_str!("Debug").to_owned(),
+    ];
+    let modules: Vec<CompiledModule> = build_framework(lib_dir)?
+        .into_iter()
+        .filter(|m| !denylist.contains(&m.self_id().name().to_owned()))
+        .collect();
+    verify_modules(&modules)?;
```
[the remark this answers is not in the document]
Comment (bmwill) [reply crates/sui-framework-build/src/lib.rs:30]: Same comment

File: crates/sui-framework/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+pub use sui_framework_build::build_move_stdlib_modules as get_move_stdlib_modules;
+pub use sui_framework_build::build_sui_framework_modules as get_sui_framework_modules;
+pub use sui_framework_build::{build_framework, build_move_package, verify_modules};
+
+// Move unit tests will halt after executing this many steps. This is a protection to avoid divergence
+const MAX_UNIT_TEST_INSTRUCTIONS: u64 = 100_000;
+
+static SUI_FRAMEWORK: Lazy<Vec<CompiledModule>> = Lazy::new(|| {
+    const SUI_FRAMEWORK_BYTES: &[u8] = include_bytes!(concat!(env!("OUT_DIR"), "/sui-framework"));
```
[the remark this answers is not in the document]
Comment (bmwill) [reply crates/sui-framework/src/lib.rs:22]: The compilation is done via the `build.rs` script which is guaranteed to be run before the library is compiled.

**Comment by bmwill** [comment]
> Ok, cool so I do not think my PR - caching genesis - is needed any more right? Since now genesis is statically written.

Yes there's a good chance your PR isn't needed anymore. It may be worthwhile for you trying just this PR out to see if things are any better

**Comment by gdanezis** [comment]
Confirmed it does resolve the test perf issue

**Pull request merged**